src/cpr_data_access/models.py

```diff
@@ -1097,3 +1105,105 @@ def to_huggingface(
         )
 
         return huggingface_dataset
+
+    def from_huggingface(
+        self,
+        huggingface_dataset: HFDataset,
+        limit: Optional[int] = None,
+    ) -> "Dataset":
+        """
+        Create a dataset from a huggingface dataset.
+
+        :param huggingface_dataset: created using `Dataset.to_huggingface()`
+        :param limit: optionally limit the number of documents to load
+        :return self: with documents loaded from huggingface dataset
+        """
+
+        document_metadata_model = self.document_model.__fields__[
+            "document_metadata"
+        ].outer_type_
+        hf_dataframe: pd.DataFrame = huggingface_dataset.to_pandas()
+
+        if limit is not None:
+            doc_ids = hf_dataframe["document_id"].unique()[:limit]
+            hf_dataframe = hf_dataframe[hf_dataframe["document_id"].isin(doc_ids)]
+
+        documents = []
+
+        for _, doc_df in tqdm(
+            hf_dataframe.groupby("document_id"),
+            total=hf_dataframe["document_id"].nunique(),
+            unit="docs",
+        ):
+            document_text_blocks = [
+                TextBlock(
+                    # TODO: we aren't able to access the original text split over lines as it's not stored in the huggingface dataset
+                    text=[row["text"]],
+                    text_block_id=row["text_block_id"],
+                    language=row["language"],
+                    type=row["type"],
+                    type_confidence=row["type_confidence"],
+                    page_number=row["page_number"],
+                    coords=[tuple(c) for c in row["coords"].tolist()] if row["coords"] is not None else None,  # type: ignore
+                )
+                for _, row in doc_df.iterrows()
+            ]
+
+            # pandas to_dict() stores sequences as numpy arrays, so we need to convert them back to lists
+            doc_fields = {
+                k: list(v) if isinstance(v, np.ndarray) else v
+                for k, v in doc_df.iloc[0].to_dict().items()
+            }
+
+            doc_metadata = document_metadata_model.parse_obj(
+                doc_fields
+                | {"source": "GST" if self.document_model == GSTDocument else "CPR"}
+            )
+
+            doc = self.document_model.parse_obj(
+                doc_fields
+                | {
+                    "document_metadata": doc_metadata,
+                    "text_blocks": document_text_blocks,
+                }
+            )
+
+            documents.append(doc)
+
+        self.documents = documents
+
+        return self
+
+    def from_hf_hub(
```

Comment:
I find the naming slightly confusing -- I think `from_hf_hub()` fits well with `from_local()` as it specifies the _location_ of where the dataset will be pulled from, whereas `from_huggingface()` is the _type_ of the input data.

Do we want to maybe make the `from_huggingface()` method private? That way the public methods exposed are consistent - and I guess we wouldn't be using `from_huggingface()` much on its own with outh the hub.

(Also I realise this is the wrong place for this comment)

Reply:
good point – i've made `from_huggingface()` -> `_from_huggingface`. I initially thought we might want to be able to load huggingface parquets directly but that's unlikely, because: a) the data uploads to HF hub will soon be automated 🤞; b) HF download deals nicely with caching; c) `from_hf_hub` could technically take a local path (i discovered this in replying to this comment).

Do we want to rename the function given the realisation it can be used locally? could just call it `from_huggingface` and call the now private function `_from_huggingface_parquet`?

Reply:
Oh that's nice, in which case I'm happy with your suggestion!